File: src/10.graph/24.241.ts
export function diffWaysToCompute(expression: string): number[] {
    // 判断 expression 是不是纯数字
    const temp = Number(expression);
    if (!isNaN(temp)) {
        return [temp];
    }

    const ans: number[] = [];

    for (let i = 0; i < expression.length; i++) {
        const ch = expression[i];
        if (!["+", "-", "*"].includes(ch)) {
            continue;
        }

        const lefts = diffWaysToCompute(expression.slice(0, i));
        const rights = diffWaysToCompute(expression.slice(i + 1));

        lefts.forEach((l) => {
            rights.forEach((r) => {
                let cur = 0;
                if (ch === "+") {
                    cur = l + r;
                } else if (ch === "-") {
                    cur = l - r;
                } else {
                    cur = l * r;
                }

                ans.push(cur);
            });
        });
    }

    return ans;
}

File: src/10.graph/24.241.go
package graph

import "strconv"

func DiffWaysToCompute(expression string) []int {
	tmp, err := strconv.Atoi(expression)
	if err == nil {
		return []int{tmp}
	}

	ans := []int{}

	for i, v := range expression {
		if v == '+' || v == '-' || v == '*' {
			lefts := DiffWaysToCompute(expression[:i])
			rights := DiffWaysToCompute(expression[i+1:])

			for _, l := range lefts {
				for _, r := range rights {
					cur := 0
					if v == '+' {
						cur = l + r
					} else if v == '-' {
						cur = l - r
					} else {
						cur = l * r
					}
					ans = append(ans, cur)
				}
			}
		}
	}

	return ans
}
